docs/DynamicCasting.md

```diff
@@ -0,0 +1,532 @@
+* Implementation: [apple/swift#29658](https://github.com/apple/swift/pull/29658)
+
+## Introduction
+
+The Swift language has three casting operators: `is`, `as?`, and `as!`.
+Each one takes an instance on the left-hand side and a type expression on the right-hand side.
+
+* The _cast test operator_ `is` tests whether a particular instance can be converted to a particular destination type.  It returns a boolean result.
+
+* The _conditional cast operator_ `as?` attempts the conversion and returns an `Optional` result: `nil` if the conversion failed, and a non-nil optional containing the result otherwise.
+
+* The _forced cast operator_ `as!` unconditionally performs the casting conversion and returns the result.  If an `as!` expression does not succeed, the implementation may terminate the program.
+
+Note: The static coercion operator `as` serves a different role and it's behavior will not be specifically discussed in this document.
+
+The following invariants relate the three casting operators:
+* Cast test: `x is T == ((x as? T) != nil)`
+* Conditional cast: `(x as? T) == (x is T) ? .some(x as! T) : nil`
+* Forced cast: `x as! T` is equivalent to `(x as? T)!`
+
+In particular, note that `is` and `as!` can be implemented in terms of `as?` and vice-versa.
+
+As with other operators with `!` in their names, `as!` is intended to only be invoked in cases where the programmer knows a priori that the cast will succeed.
+If the conversion would not succeed, then the behavior may not be fully deterministic.
+See the discussion of Array casting below for a specific case where this non-determinism can be important.
+
+The following sections detail the rules that govern the casting operations for particular Swift types.
+Except as described below, casting between dissimilar types always fails -- for example, casting a struct instance to a function type or vice-versa.
+
+Where possible, each section includes machine-verifiable _invariants_ that can be used as the basis for developing a robust test suite for this functionality.
+
+## Identity Cast
+
+Casting an instance of a type to its own type will always succeed and return the original value unchanged.
+
+```
+let a: Int = 7
+a is Int // true
+a as? Int // Succeeds
+a as! Int == a // true
+```
+
+## Classes
+
+Casting among class types follows standard object-oriented programming conventions:
+
+* Class upcasts: If `C` is a subclass of `SuperC` and `c` is an instance of `C`, then `c is SuperC == true` and `(c as? SuperC) != nil`.  These "upcasts" do not change the representation in any way.  However, when `c` is accessed via a variable or expression of type `SuperC`, only methods and instance variables defined on `SuperC` are available.
+
+* Class downcasts: If `C` is a subclass of `SuperC` and `sc` is an instance of `SuperC`, then `sc is C` will be true iff `sc` is actually an instance of `C`.  When this "downcast" does succeed, it does not change the representation.
```

Comment:
There needs to be some mention of the runtime/ABI representation of a value, and how type coherence applies before this and the one above it can mention its effect. The goal is to state that casts never "slice" or "extend" instances at runtime, right?

Reply:
As far as possible, I'd like to specify the surface language behavior independently of the details of the runtime/ABI representation.  Referring to "representation change" here is admittedly rather sloppy.  Is there a better way to communicate the idea that class casts do not change the referenced value?

Reply:
I've rephrased slightly to hopefully make this clearer.  If you have better ideas, please let me know.